### DISCUSSION TITLE
[RFC] Indexer API

### DISCUSSION THREAD

**Discussion by shilman** [storybookjs/storybook#23176] [RFC]
**Status: accepted**
**Tracked in: https://github.com/storybookjs/storybook/issues/23457**

### Summary

This proposes a new API for indexing stories and documents in Storybook. This makes it possible to support alternative syntaxes for stories and documents. For example, if somebody from the community wanted to support story authoring as Vue SFCs or docs in Markdown or Markdoc.

### Problem Statement

[Component Story Format (CSF)](https://storybook.js.org/docs/react/api/csf) is a portable standard for capturing UI component examples. It is based on [ECMAScript Modules (ESM)](https://tc39.es/ecma262/#sec-modules), where the default export contains metadata about the component being documented, and each named export captures a key state of that component.

CSF has been incredibly successful in Storybook and has even been adopted by other tools in the frontend ecosystem. However, it has some key criticisms as well:

1. **Static.** Because ESM named exports are statically defined, it is not easy to dynamically generate stories in CSF. This make it hard to do combinatorial testing on component props, or automatically generate stories from static fixture data (both of which were fairly easy to do with the legacy [storiesOf API](https://github.com/storybookjs/storybook/blob/next/code/lib/preview-api/docs/storiesOf.md) that CSF replaced).
2. **Fixed syntax.** Renderers like Vue and Svelte have their own syntaxes for defining and using components, and some users would prefer to write their stories using these alternative syntaxes. Similarly, Storybook supports MDX for long-form documentation, but in the future might want to support pure [Markdown](https://daringfireball.net/projects/markdown/syntax) or [Markdoc](https://markdoc.dev/) instead.

We want an API that makes it possible to solve these problems in a flexible way.

### Non-goals

1. **Not a storiesOf replacement.** We are not trying to solve `storiesOf` backwards compatibility. We are betting on static analysis and the solution described here is a “load-time” solution. We may provide a more dynamic solution in the future for dynamic “run-time” generating variants *within* a story and handling them sensibly across our tools.
2. **Not an addon.** We are not building addons for fixture-based stories, combinatorial stories, or alternative story syntaxes. We may do so in the future, but we want to make it possible for the community to experiment with options.
3. **No loader/plugin tools.** The solution described here assumes that the user’s story files are importable as ESM, which may require a “load-time” transformation such as a Vite plugin or Webpack loader. This proposals does not try to solve that problem, though our documentation will include examples for how to do this.

### Implementation

To solve these problems, we propose a public indexer API for Storybook. The API would make it possible to extend Storybook in a variety of ways.

#### Existing architecture

Before introducing the API, let’s review Storybook’s architecture:

**Config.** Users define the files to index in `.storybook/main.js`. (The field is called `stories`, but should be renamed to `index` since it can contain docs as well, and potentially other types of entries in the future):

```jsx
export default {
  stories: ['../src/**/*.mdx', '../src/**/*.stories.js'],
};
```

**Index.** The index is a global listing of all the entries in your storybook. It’s an in-memory data structure that gets updated dynamically as the user updates the file system, and is also served at the URL endpoint `/index.json` in v7 storybooks. It also defines what appears in Storybook’s sidebar.

Each file results in one or more entries in the index and in the sidebar. For example, consider this button example and its corresponding `index.json` entry and sidebar visualization.

```jsx
// Button.stories.js
import { Button } from './Button';

export default {
  title: 'Example/Button',
  component: Button
};
export const Primary = { args: { primary: true } };
export const Secondary = { };
```

```json
// index.json
{
  "example-button--primary": {
    "id": "example-button--primary",
    "title": "Example/Button",
    "name": "Primary",
    "importPath": "./src/stories/Button.stories.ts",
    "tags": ["story"],
    "type": "story"
  },
  "example-button--secondary": {
    "id": "example-button--secondary",
    "title": "Example/Button",
    "name": "Secondary",
    "importPath": "./src/stories/Button.stories.ts",
    "tags": ["story"],
    "type": "story"
  }
}
```
<img width="433" alt="indexer-figure" src="https://github.com/storybookjs/storybook/assets/488689/1a978c24-0c6b-4c0a-b992-4f014ed45cf5">

#### Implementation

The indexer is a function that generates zero or more index entries from an input file.

```ts
// type ExportName/Path/ComponentTitle/StoryName/StoryId/Tag = string;

export interface IndexerOptions {
  makeTitle: (userTitle?: string) => string;
}

/**
 * An indexer describes which filenames it handles, and how to index each individual file - turning it into an entry in the index.
 */
export type Indexer = {
  /**
   * A regular expression that should match all files to be handled by this indexer
   */
  test: RegExp;
  /**
   * Indexes a file containing stories or docs.
   * @param fileName The name of the file to index.
   * @param options {@link IndexerOptions} for indexing the file.
   * @returns A promise that resolves to an array of {@link IndexInput} objects.
   */
  createIndex: (fileName: string, options: IndexerOptions) => Promise<IndexInput[]>;
};

/**
 * The base input for indexing a story or docs entry.
 */
export type IndexInput = {
  type: 'story';
  /** The file to import from e.g. the story file. */
  importPath: Path;
  /** The name of the export to import. */
  exportName: ExportName;
  /** The name of the entry, auto-generated from {@link exportName} if unspecified. */
  name?: StoryName;
  /** The location in the sidebar, auto-generated from {@link importPath} if unspecified. */
  title?: ComponentTitle;
  /**
   * The custom id optionally set at `meta.id` if it needs to differ from the id generated via {@link title}.
   * If unspecified, the meta id will be auto-generated from {@link title}.
   * If specified, the meta in the CSF file _must_ have a matching id set at `meta.id`, to be correctly matched.
   */
  metaId?: MetaId;
  /** Tags for filtering entries in Storybook and its tools. */
  tags?: Tag[];
  /**
   * The id of the entry, auto-generated from {@link title}/{@link metaId} and {@link exportName} if unspecified.
   * If specified, the story in the CSF file _must_ have a matching id set at `parameters.__id`, to be correctly matched.
   * Only use this if you need to override the auto-generated id.
   */
  __id?: StoryId;
};

```

Note that this API only covers how to add `story`/`docs` entries to the index. It does not specify how to load those files. For that, your addon or preset would need to provide some other kind of configuration, such as a webpack/vite plugin, to load the `importPath` using whatever builder the user has configured in their project. For example, if you are creating an addon for writing stories as Vue SFCs, the indexer will tell Storybook the list of stories available in each Vue SFC story file, and the addon will provide a vite and webpack plugin (possibly using [unplugin](https://github.com/unjs/unplugin) for convenience) to transform those SFCs into CSF at load time. We will elaborate on this more as needed.

Users can register the indexers in their `.storybook/main.js` config file and addons can register indexers using the [presets API](https://storybook.js.org/docs/react/addons/writing-presets) (which is the same as `main.js`).

```ts
interface StorybookConfig {
  indexers: Indexer[] | async (existing: Indexer[], options: Options) => Promise<Indexer[]>
}
```

##### Fixture data

Consider the following addon that generates stories for components based on JSON fixture data. It looks for `jsonstories.js` files in the project and then for each such file, looks for `.json` files in the same directory, each of which is added as a story.

```ts
import fs from 'fs/promises';
import path from 'path';

const JsonIndexer: Indexer = {
  test: /jsonstories.[tj]sx?$/,
  createIndex: async (fileName) => {
    // JSON files in the current directory
    const jsonFiles = (await fs.readdir(path.dirname(fileName))
      .filter(f => f.endsWith('.json'))

    // Sidebar entries for each file
    return jsonFiles
      .map(jsonFile => ({
        type: 'story',
        importPath: fileName,
        exportName: f.split('.')[0]
      });
  }
}

export default { indexers: [JsonIndexer] };
```

##### Combinatorial testing

Consider the following addon that supports combinatorial testing via an extra parameter `combos` in the default export which specifies the arg values to permute. It generates extra named exports of the form `Combo0 ... ComboN` for each combination of args.

```ts
import fs from 'fs/promises';
import path from 'path';

const CombosIndexer: Indexer = {
  test: /.combos.[tj]sx?$/,
  createIndex: async (fileName) => {
    const code = await fs.readFile(fileName);
    const combos = parseCombos(code);

    // Sidebar entries for each file
    return combos.map((combo, idx) => ({
      type: 'story'
      importPath: fileName,
      key: `Combo${idx}`
    }];
  }
}

export default { indexers: [MarkdownIndexer] };
```

##### Markdown docs

Consider the following addon that supports pure Markdown docs and uses the headmatter to optionally control the sidebar:

```tsx
import fs from 'fs/promises';
import path from 'path';

const MarkdownIndexer: Indexer = {
  test: /.md$/,
  createIndex: async (fileName) => {
    const md = await fs.readFile(fileName);
    const { title, name, tags } = parseHeadMatter(md);

    // Sidebar entries for each file
    return [{
      type: 'docs'
      importPath: fileName,
      exportName: 'default',
      title, // optional
      name, // optional
      tags, // optional
      storiesImports: undefined // not needed since we don't ref stories
    }];
  }
}

export default { indexers: [MarkdownIndexer] };
```

### Prior Art

There is already an experimental indexer API in Storybook, it’s just not publicly documented. The primary difference in this proposal:

1. The return type of the current API is a `CSFFile` object, which is an implementation detail. The new API is a simple interface that doesn’t dictate an implementation.
2. The existing API does not allow for pluggable docs implementations. This proposal adds one.
3. The existing API includes a preset called `storyIndexers`. The new API is `indexers` to reflect that it can also produce `docs` entries.

### Deliverables

- [ ]  Implement the API and expose it as part of the Storybook type system
- [ ]  Convert the existing indexers to the new API
- [ ]  Deprecate or remove the old API. Removing would be a breaking change, but since it’s an experimental API we are free to do that.
- [ ]  Document the API in our addons documentation with examples (as above)
- [ ]  A release post describing the API, its existing usage (Svelte CSF, MDX stories, pure Markdown docs), and potential future applications

### Risks

1.  **Unsupported use cases.** One risk is that there are use cases we should be considering that we're not. We release the API and have to make disruptive breaking changes later to accommodate those use cases.
2. **Over-flexible API.** Another risk is that the API is somehow too flexible, and supporting it publicly somehow restricts future Storybook architectural decisions.

### Unresolved Questions

- [ ]  Should we think about other sidebar entry types as part of this proposal? External links?
- [ ]  Should we think about future sidebar UIs, e.g. a “docs” sidebar where each navigation point under a page is a heading in the longform docs?
- [ ]  Should we consider utilities for writing CSF to make it easier to write loaders/plugins? Our current `CSFFile` does this, but is not ergonomic and we wouldn’t want to commit to this as a public API.
- [ ]  Should we deprecate the existing internal API or just release this API as a breaking internal change?

**Comment by tobiasdiez** [comment]
This sounds like a very good proposal! I also like that it will be possible to index docs and not only stories. As far as I can see, it fulfills all requirements I have on it in the context of [native Vue SFC](https://github.com/tobiasdiez/storybook-vue-addon). If you like, I can test early prototypes of the indexer api in this setting.

Concerning the half-sentence, "the addon will provide a vite and webpack plugin to transform those SFCs into CSF at load time" - why is it still necessary to generate CSF? Wouldn't it be better/more flexible/... if the second step would be a loader that returns the story object for a given storyid, without necessarily going through CSF. I understand that this may be more related to https://github.com/storybookjs/storybook/discussions/23177, then the Indexer API in itself, but it is one of the main pain points I've encountered while working on the vue addon.

**Reply by shilman** [reply]
@tobiasdiez thanks so much for offering to test this when it's ready! i'll def keep you posted.

re: generating CSF, that's a great question. i didn't think about it too hard to be honest. can you propose what the "story ID to story object" format would look like, if it wasn't CSF? i'd be open to considering alternatives.

**Reply by tobiasdiez** [reply]
The problem in the Vue SFC integration is that I load the vue file and parse it via the vue compiler. At this stage, I would have already enough information in memory to render the story. But storybook expects a CSF, so I have to serialize the vue story into CSF, which is then parsed/loaded again by storybook. So it would be nice if one could simply cut out the middle man and not pass through CSF at all. 

As another application of this, imagine you generate your stories from dynamic infos retrieved via an expensive API. The indexer would simply call the API to get a list of stories, and the loader would ideally only need to get the information for the particular story that is rendered, and not for every possible one (to generate a full CSF file).

So in short the loader would return a `StoryFn` for a given story id, in whatever form.

**Reply by shilman** [reply]
The problem is that there are two different places where we need the story information:
1. When we generate the index and serve it on a route
2. In a webpack/vite loader where it gets bundled up for the browser

So this RFC is really about defining the data format for #1 that is hopefully simple & somewhat future-proof (famous last words).

The browser bundling part is more complicated. Storybook's code dynamically imports a module based on the relative story path (e.g. `./src/stories/Button.stories.vue`). It doesn't really care about the details of the file, but it expects to see a default export, possibly with a `title`/`component`/etc. and a bunch of named exports. Currently the format of that file can actually be CSF1/2/3 -- the code in Storybook core normalizes it. But it must be ESM, so your compiler must output ESM of some form, which I'm sure Vue compilers are set it up to do.

There's a whole other version of the world where you can run the entire "preview iframe" yourself, in which case you're free to do whatever you want. However, the code in Storybook for managing decorators, HMR, manager/preview synchronization, etc. is pretty complicated, so we don't recommend that route.

**Reply by tobiasdiez** [reply]
Yes, you are right, what I mentioned concerns more the second point and not the first (the actual subject of this RFC). So sorry for the off-topic.

My idea was to make the second point a public interface as well, similar to what the indexer api provides for the first point.
Say
```ts
interface Loader {
  getMetaInfo: (string storypath) => Meta 
  getStory: (string storypath, string storyid) => StoryFn
}
```
which could be (roughly) implemented as a CSF loader to load CSF files (only the getStory function for simplicity):
```ts
getStory(string storypath, string storyid) {
    storyFile = import(storypath) // This uses just a normal ts/js loader
    return storyFile[storyid]
}
```
or vue/svelte stories:
```ts
getStory(string storypath, string storyid) {
    storyFile = import(storypath) // storypath is an ordinary vue file, so this now uses the standard vue file loader
    return extractStory(storyid) // this is now more involved, and needs to convert the vue object to a proper StoryFn with render function etc
    // note that this doesn't use the CSF loader
}
```
or API calls:
```ts
getStory(string storypath, string storyid) {
    story = callApi(storyPath, storyid) // no import at all
    return convert(story) // convert the api result to a proper StoryFn with render function etc
}
```

**Reply by shilman** [reply]
Aha that's super interesting. Cc @tmeasday

**Reply by tmeasday** [reply]
If I understand correctly, we are really taking about a runtime vs build time transformation. 

A **loader** (build time transformer) would take in a string that represents a vue/svelte story and output a string that represents CSF.

A *runtime transformer* would basically be a wrapper around the called to `importFn(importPath)`:

https://github.com/storybookjs/storybook/blob/6d0596f8b2462e6e0ae9d9dd72792e361c437878/code/lib/preview-api/src/modules/store/StoryStore.ts#L150

You could actually implement this right now by redefining the `importFn()` itself [1]. That's defined by the builder right now so it's not really overridable (but you could prototype by changing the builder). We could definitely consider an API where there's a similar concept to an indexer, like a transformer maybe?

```ts
const JsonTransformer: Transformer = {
  test: /jsonstories.[tj]sx?$/,
  transform: (jsonExports) => {
    /* do stuff */
    return csfExports;
  }
}
```

Does that makes sense @tobiasdiez? I think implementation-wise that'd be easy for us to do (we'd just need to support it in each builder). My only worry would be performance regressions if the runtime transformer is slow.


[1] the api is just `(path: Path) => Promise<ModuleExports>`, where `ModuleExports` needs to look a like a CSF file, so not the exact API you outlined above, but pretty close IMO.

**Reply by shilman** [reply]
I like this idea @tmeasday. It's probably easier in most cases to make these transformations at runtime than at build time. Let's see what @tobiasdiez say about Vue and maybe @JReinhold @j3rem1e say on the Svelte side. JSON fixture data would also be pretty trivial this way. 🤔

**Reply by tobiasdiez** [reply]
Yes, this is a nice idea. For the performance, one could try to put as much of the heavy stuff in the build-time loader and only do a bit of light post processing in the transformer.

**Reply by JReinhold** [reply]
These are all interesting ideas.
Is there something about Node vs. browser we need to consider here? The indexer runs in Node, but I would expect a runtime transformer would need to run in the browser (I might be wrong), which could impose limitations on what's doable and which packages can be used. And how would this work for static builds?

**Reply by tmeasday** [reply]
Yes, the transformer would run in the browser (both for dev + built storybooks), so the definition would probably have to reference a separate file, like:

```js
const JsonTransformer: Transformer = {
  test: /jsonstories.[tj]sx?$/,
  transformerPath: './transformer.js',
}
```

**Reply by j3rem1e** [reply]
I'm not sure to understand everything here, but if the transformer function should run in the browsr, it will not be optimal for Svelte file. The svelte compiler is designed to run on the backend. It is possible to ship it in the browser (like Svelte REPL) but it's slow

**Comment by thibaudcolas** [comment]
> or automatically generate stories from static fixture data (both of which were fairly easy to do with the legacy [storiesOf API](https://github.com/storybookjs/storybook/blob/next/code/lib/preview-api/docs/storiesOf.md) that CSF replaced)

How about generating stories from _dynamic_ fixture data (say an API call)? I’ve been hoping to do this for a while, it would simplify my integration a _lot_, but I’ve always put this off due to the uncertainty about `storiesOf`. Just wondering if there is _any_ room for this in this Indexer API or other ways to add entries to the index "later" in the build process.

edit: I see #23177 does specifically say _Fixture data. Some users have fixture data in JSON, YAML, databases, or API servers and want to automatically turn this data into stories._. So I take it that this would still be possible, but only from API calls at build time? (since the Indexer API has async support)

> ## Non-goals
> **Not a storiesOf replacement.** We are not trying to solve storiesOf backwards compatibility. We are betting on static analysis and the solution described here is a “load-time” solution. We may provide a more dynamic solution in the future for dynamic “run-time” generating variants within a story and handling them sensibly across our tools.

I’m concerned to see this called out with this language, as to me it seems #23177 clearly states this Indexer API is intended as a replacement for people currently stuck with `storiesOf` for other reasons than inertia.

> is also served at the URL endpoint /index.json in v7 storybooks. 

This is the first time I see this formally described so apologies if this is obvious – is the presence of this index representation and its structure something that’s part of the stable API, that other tools could rely on? Again from my perspective for the purpose of integration with other pattern library tools, it’d be really useful to have control over adding entries to Storybook’s index (what this RFC solves), but also at other points inspecting what stories are defined in Storybook according to its parsing of CSF and Indexer API calls that happened at build time.

**Reply by shilman** [reply]
> How about generating stories from dynamic fixture data (say an API call)?

We generally don't like dynamic fixture data in Storybook since it means that your components are not really isolated. I would recommend considering carefully if that's what you really want to do.

That said, we know people want to do this and Storybook supports it in two different ways.
1. The [dynamic stories POC](https://stackblitz.com/edit/github-h2rgfk?file=README.md) linked from the [StoriesOf Deprecation RFC](https://github.com/storybookjs/storybook/discussions/23177) is capable of fetching data and generating stories from it.
2. The [Storybook loaders API](https://storybook.js.org/docs/react/writing-stories/loaders) allows you to asynchronously load data before your story renders.

> I’m concerned to see this called out with this language, as to me it seems https://github.com/storybookjs/storybook/discussions/23177 clearly states this Indexer API is intended as a replacement for people currently stuck with storiesOf for other reasons than inertia.

This is all worded pretty carefully, but I guess it's not clear enough. The Indexer API is a kind of "meta-replacement". I.e. it's not something you can use directly, but you can use it to build stuff like the [dynamic stories POC](https://stackblitz.com/edit/github-h2rgfk?file=README.md). We still haven't decided exactly how or whether we want to create a direct replacement for `storiesOf`, e.g. the Variants ideas mentioned at the end of the [StoriesOf Deprecation RFC](https://github.com/storybookjs/storybook/discussions/23177).

> This is the first time I see this formally described so apologies if this is obvious – is the presence of this index representation and its structure something that’s part of the stable API, that other tools could rely on? 

Exactly. We already have an indexer API but it's not really documented. The intent of this RFC is to make it public & documented & supported so that people can build tools on top of it.

**Comment by tobiasdiez** [comment]
One more small remark: while implementing the indexer for the vue stories, I encountered the problem how to calculate the id of a story (based on the story name and some meta info), see https://github.com/tobiasdiez/storybook-vue-addon/pull/38#issuecomment-1464930276. If I read the proposal correctly, the indexer doesn't need to do this anymore and this is very good.

**Comment by floroz** [comment]
Thanks for your work on this proposal.

I am trying to play around with the new Indexer API but I am confused about its usage (I know you are working on providing some documentation soon).

I am using the addon manager API with the new experimental config

```ts
// export of the addon for clients to add to their `addons: []` in `.storybook/main.ts`

const vueSFCIndexer: Indexer = {
  test: /\.stories\.vue$/,
  index,
}

export const experimental_indexers: StorybookConfig['experimental_indexers'] = existingIndexers => [
  vueSFCIndexer,
].concat(existingIndexers || [])
```

The `index` function is the following:

```ts

export async function index(
  fileName: string,
  { makeTitle }: IndexerOptions,
): Promise<IndexInput[]> {
  const code = fs.readFileSync(fileName, { encoding: 'utf-8' }).toString()

  const { meta, stories } = parseSFCStory(code)

  return stories.map(({ title }) => ({
    type: 'story',
    importPath: fileName,
    exportName: title,
    title: makeTitle(meta.title),
  }))
}
```

If I am understanding correctly the `IndexInput` type signature

- `importPath` is going to reference the original filename containing a pre-transformed Story. 

Will be up to our custom loader to make sure that Story will be transformed to CSF format (in this case the filename is pointing at `*.stories.vue` files.
This `importPath` will be parsed again at story load time where the actual file will be transformed. What is its purpose, since it was not included in the deprecated `storyIndexer` API?

- `exportName` is the export name of a single story (can be generated by the individual story title)

- `title` the title of the `meta` used to group together in the navigation panel the different Story `exportName`/`name`


So for example, if my `meta.title` is `Components/Button`, then each `exportName` can be _Primary_, _Secondary_ etc.

Is my understanding correct?

**Comment by kamranayub** [comment]
I just wanted to chime in, and I apologize if this isn't exactly what the Indexer API is meant to support (but it sounds closest to what I want). I haven't done a wide search yet on other issues/discussions. 

I am currently hacking Storybook into the Docusaurus lifecycle as a plugin.

You can see it here as a WIP:

https://github.com/excaliburjs/Excalibur/commit/b42edc1444f3464ec6651b603aa68dc2a8cc9ab0

Currently it's not ideal: I have to basically run the Storybook dev server side by side while Docusaurus is running and then embed the dev server URL to iframe stories into our docs. For a static build, it's a little easier, it's mainly dev that is sub-optimal. But in the end, we still are getting a separate "site" that isn't natively integrated with our own. 

What I "want" is a "headless" Storybook API so I can create a native Docusaurus integration. That's the long and short of it.

For example, we use Typedoc and it's relatively easy to call their Node API and get a JSON data structure you can use to render your own UI. It's more work in userland but you get a native integration. That's exactly how the Docusaurus Typedoc plugin works and it's awesome. 

But AFAIK, before seeing this Indexing RFC, there's no way to call a core API and get all the data we'd need back to render a custom UI. Even the RFC specifies here that we'd need to still go through a builder/loader to get story metadata.

This is the first time I have seen this but I did notice the indexing code when I was trying to figure this out last night, so I will be diving into this more. Perhaps what's there is enough I can piece together something, since the Indexer API gives me the import path which may be just enough + the builder.

I'm happy to move this to a dedicated discussion but this _seems_ like the type of use case this is meant to (at least partially) address.

**Comment by psychobolt** [comment]
I think currently the `storyStoreV7` builder does not take account of the indexer's `importPaths`. I tried creating my own `indexer` following the experimental doc guide, using a different `importPath` from `fileName`. However, my dynamically generated `importPath` is not included in the virtual module `storybook-stories.js`. Another thing to consider is transforming these new CSF files for `test-runner`

**Reply by JReinhold** [reply]
Sounds like a bug, can you open an issue with a minimal reproduction?

**Reply by psychobolt** [reply]
A issue was created by someone else: https://github.com/storybookjs/storybook/issues/25554

**Comment by DaniGuardiola** [comment]
I see this is closed, but [the docs](https://storybook.js.org/docs/api/main-config/main-config-indexers?utm_source=chatgpt.com) still refer to it as experimental:

<img width="664" height="228" alt="image" src="https://github.com/user-attachments/assets/4902a37b-33fc-4087-918e-1cc81b1d0e01" />

So, it is safe to use? Or will it be removed/changed in the future?
